kAFL-Fuzzer/common/qemu.py

```diff
@@ -425,55 +410,99 @@ def shutdown(self):
         except:
             pass
 
-        return self.process.returncode
+        for tmp_file in [
+            self.payload_filename,
+            self.tracedump_filename,
+            self.control_filename,
+            # self.binary_filename,
+            self.bitmap_filename,
+        ]:
+            try:
+                os.remove(tmp_file)
+            except FileNotFoundError:
+                pass
+
+        retcode = self.process.returncode
+        self.process = None
+        return retcode
 
     def __set_agent(self):
-        agent_bin = self.config.argument_values['agent']
-        bin = read_binary_file(agent_bin)
-        assert (len(bin) <= self.agent_size)
-        atomic_write(self.binary_filename, bin)
+        agent_bin = self.config.argument_values.get('agent', None)
+        if agent_bin is not None:
+            bin_ = read_binary_file(agent_bin)
+        else:
+            bin_ = b''
+        assert (len(bin_) <= self.agent_size)
+        try:
+            with open(self.binary_filename, 'xb') as f:
+                f.truncate(self.agent_size)
+                f.seek(0)
+                f.write(bin_)
+        except FileExistsError:
+            pass
 
     def start(self):
 
         if self.exiting:
             return False
 
-        self.persistent_runs = 0
-        self.handshake_stage_1 = True
-        self.handshake_stage_2 = True
+        if self.process is not None:
+            print_fail("Trying to launch Qemu second time?")
+            raise Exception
 
-        if self.qemu_id == "0" or self.qemu_id == "1337": ## 1337 is debug instance!
-            log_qemu("Launching virtual machine...CMD:\n" + ' '.join(self.cmd), self.qemu_id)
-        else:
-            log_qemu("Launching virtual machine...", self.qemu_id)
+        max_attempts = 10
+        for attempt in range(max_attempts):
```

Comment:
why start looping here? If Popen() fails we probably have a fatal error like broken cmdline. Retry / waiting would make more sense starting with the qemu_connect and qemu_handshake steps.

Reply:
Hmm, on second thought, these attempts were only needed before the fix.